enable to recursively copy directories to remote

jeanguyomarch opened the pull request:
Hi,

it seems this repository does not have recent activity, but it found it quite useful (thanks :))
I had trouble with copying **directories** from the host to the remote, and it seems this is because of how `machinectl` handles existing files with `copy-file`.
It took me some time to pinpoint the problem and find a "solution" (that's more a kludge than a "real" solution, but anyway...). I hope this can be of some use.

----

Machinectl cannot overwrite existing files with `copy-to`.  There is a
pending RFE (systemd/systemd#9441), but currently no other alternative
(to the best of my knowledge) than proactively remove the existing file
before pushing a new one.

This is especially annoying when copying a whole directory. As this is
done recursively file by file, and since the ansible script pushed on
the remote has the same path, this currently triggers machinectl which
rejects the copy with a "file exists" message.

The kludge proposed in this path is to systematically attempt to
remove existing files before pushing them to the remote... That's
definitely not great, but I have not found better...

Commit:
enable to recursively copy directories to remote

Machinectl cannot overwrite existing files with `copy-to`.  There is a
pending RFE (systemd/systemd#9441), but currently no other alternative
(to the best of my knowledge) than proactively remove the existing file
before pushing a new one.

This is especially annoying when copying a whole directory. As this is
done recursively file by file, and since the ansible script pushed on
the remote has the same path, this currently triggers machinectl which
rejects the copy with a "file exists" message.

The kludge proposed in this path is to systematically attempt to
remove existing files before pushing them to the remote... That's
definitely not great, but I have not found better...

tomeon commented:
Hi @jeanguyomarch -- thank you very much for your contribution!  Glad that you've found this connection plugin useful.

Would you be willing to add some tests for this scenario?  A regression test for the `File exists` error would be particularly helpful.  I'd also like to verify that this change doesn't, for example, result in the `copy` module clobbering files on the target machine even when the `force` argument is `false`.  I assume this is unlikely, but would like to make sure.

I look forward to merging your changes :)

tomeon commented on connection_plugins/machinectl.py:
```diff
@@ -574,13 +574,27 @@ def put_file(self, in_path, out_path):
+        # on the remote target) must be overwritten for each file in the
+        # directory.
+        # Without removing the file first, we get an error: "file exists".
+        returncode, stdout, stderr = self._run_command('shell', args=['/bin/sh', '-c', f'rm -rf "{out_path}"'], machine=self.machine)
```
Although it's probably safe to assume that a `systemd-machined` guest is Linux (or at least POSIX-compliant), I would prefer something like the following:

```
remove_cmd = self._shell.remove(out_path, recurse=True)
returncode, stdout, stderr = self._run_command('shell', args=remove_cmd, machine=self.machine)
```
This respects customization the user has done with respect to shell selection (e.g. setting [`ansible_shell_executable`](https://docs.ansible.com/ansible/latest/user_guide/intro_inventory.html#ansible-shell-executable)).

Additionally, `_shell.remove` handles quoting the `path` argument, if necessary.

tomeon replied:
Hmm, darn.  Couple of subtleties/hurdles:

1. `self._shell.remove` returns a command string; needs to be run through `shlex.split` before being run through `self._run_command`.
2. `self._shell.remove` returns an unqualified path for `rm`, and `machinectl shell` requires an absolute path.

Thinking a combination of these approaches is maybe the way to go:

```
remove_cmd = self._shell.remove(out_path, recurse=True)
remove_args = [self._play_context.executable, '-c', remove_cmd]
returncode, stdout, stderr = self._run_command('shell', args=remove_args, machine=self.machine)
```

Though this assumes that `self._play_context.executable` recognizes the `-c` option.  

Thoughts?

jeanguyomarch replied:
To the best of my knowledge, there is no support for systemd-* on non-linux OSes (that are still maintained). 
But I agree that trying to embrace Ansible's abstraction as much as possible is desirable. I'll update my code.

There is indeed a problem with the path for `rm`, I haven't found better than your solution: ask for the shell (managed by Ansible) to execute the command.  I can see in [Ansible's code](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/shell/sh.py) that `sh` shall be POSIX-compliant; fortunately it seems that `-c` is [standardized and required](https://pubs.opengroup.org/onlinepubs/9699919799/).
I guess this is a reasonable assumption.

tomeon commented on connection_plugins/machinectl.py:
```diff
@@ -574,13 +574,27 @@ def put_file(self, in_path, out_path):
 
     def fetch_file(self, in_path, out_path):
-        super(Connection, self).put_file(in_path, out_path)
+        super(Connection, self).fetch_file(in_path, out_path)
```
Whoops :grimacing: -- nice catch!

tomeon commented on connection_plugins/machinectl.py:
```diff
@@ -574,13 +574,27 @@ def put_file(self, in_path, out_path):
+        # Without removing the file first, we get an error: "file exists".
+        returncode, stdout, stderr = self._run_command('shell', args=['/bin/sh', '-c', f'rm -rf "{out_path}"'], machine=self.machine)
+        if returncode != 0:
+            raise AnsibleError('failed to perform cleanup of file {0}:\n{2}\n{3}'.format(out_path, stdout, stderr))
```
I think some of the format placeholders are misnumbered; shouldn't it be the following?:
```
raise AnsibleError('failed to perform cleanup of file {0}:\n{1}\n{2}'.format(out_path, stdout, stderr))
```

jeanguyomarch replied:
oops, yes indeed

jeanguyomarch commented:
I'll add tests as requested. I've just noticed there is [a travis-ci integration](https://travis-ci.org/github/tomeon/ansible-connection-machinectl/jobs/744745785), and my PR seems to make the existing test suite fails (when removing a file)...
Since the previous build (master) was successful, I guess this has something to do with this PR.
As I'm a bit reluctant to install Vagrant and Docker on my own machine, I'll hijack travis for testing ;)

jeanguyomarch commented:
So.. I've updated  my initial commit with additional tests... but the build still fails on [travis-ci](https://travis-ci.org/github/tomeon/ansible-connection-machinectl/builds/747798817).
This is quite annoying, as running similar tests locally "works as expected". Maybe this is due to different ansible versions ? (I'm using 2.10.3). I'm not sure what to look for.

tomeon committed:
Add Ansible 2.{8,9,10} and Python 3.{7,8}

tomeon committed:
run-tests: create dummy TTY as needed

i.e., if

    1. the "tty" command exits with a nonzero status and
    2. Reading a zero-length string from stdin exits with nonzero
       status.

For as-yet-unclear reasons, certain tests fail if stdin is closed or
connected to a regular file (or pipe, or any non-tty file?).  The
run-tests script now sets up a TTY on stdin using socat, which allows
the affected tests to pass.

tomeon commented:
Think I've partially tracked down the source of the test failures.  It looks like **something** (not sure if it's Docker, Tox, Ansible, `machinectl`, this connection plugin, or something else) doesn't like it when stdin is something other than a TTY.  I'm not sure why this issue manifested with your changes, but you can see from [these Travis CI results](https://travis-ci.org/github/tomeon/ansible-connection-machinectl/builds/747964717) that [setting up a TTY on stdin before running `tox`](https://github.com/tomeon/ansible-connection-machinectl/commit/23102e9f41c6453652c19bb86161ef72ac6ee9dc) allows `ansible-playbook` to succeed in running the `test.yml` playbook (note that the Travis jobs are still failing, but `ansible-playbook` succeeds... something more to dig into).

I'll post here again once I've figured out why the Travis builds fail; then perhaps you could merge or cherry-pick my changes into your branch?

tomeon commented:
[Newest Travis builds are succeeding](https://travis-ci.org/github/tomeon/ansible-connection-machinectl/builds/747978788).

jeanguyomarch commented:
Thanks for taking the time to investigate the problem. I've squashed my commits and cherry-picked yours on top.

tomeon commented:
Merged!  TYVM, @jeanguyomarch :)
